pkg/provider/consulcatalog/config.go

```diff
@@ -85,6 +102,60 @@ func (p *Provider) buildConfiguration(ctx context.Context, items []itemData) *dy
 	return provider.Merge(ctx, configurations)
 }
 
+func connectTransportName(n string) string {
+	return "connect-tls-" + n
+}
+
+type connectCert struct {
+	service string
+	root    []string
+	leaf    keyPair
+}
+
+func (c *connectCert) getRoot() []tls.FileOrContent {
+	var result []tls.FileOrContent
+	for _, r := range c.root {
+		result = append(result, tls.FileOrContent(r))
+	}
+	return result
+}
+
+func (c *connectCert) getLeaf() tls.Certificate {
+	return tls.Certificate{
+		CertFile: tls.FileOrContent(c.leaf.cert),
+		KeyFile:  tls.FileOrContent(c.leaf.key),
+	}
+}
+
+func (c *connectCert) serverTransport(item itemData) *dynamic.ServersTransport {
+	return &dynamic.ServersTransport{
+		ServerName:         connectTransportName(item.Name),
+		InsecureSkipVerify: true,
+		RootCAs:            c.getRoot(),
+		Certificates: tls.Certificates{
+			c.getLeaf(),
+		},
+		VerifyPeerCertificate: func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			// We should use RootCAs here, but consul expect that in ClientCAs (don't ask)
+			// https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L279-L282
+			// called via https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L258
+			cfg.ClientCAs = cfg.RootCAs
+			cert, err := verifyChain(cfg, rawCerts, true)
+			if err != nil {
+				return err
+			}
+			certs := []*x509.Certificate{0: cert}
+			uri := &connect.SpiffeIDService{
+				Host:       item.Node,
```

Comment:
If I compare with the equivalent code from the consul connect package, (github.com/hashicorp/consul/connect/resolver.go) it seems to me that the Host is actually never set. Which does not result in any problem because during the URI comparison (in verifyServerCertMatchesURI), the Host is overwritten anyway with the one from the cert's URI. So I think to be as close as possible, we should not set Host here, right?

Reply:
Jupp, also removed the `ServerName` above since it is not used. Let's see what the tests say :)